packages/sanity/src/structure/panes/documentList/useDocumentSheetColumns.tsx

```diff
@@ -44,43 +51,93 @@ const PreviewCell = (props: {
   )
 }
 
-const TableTextInput = (props: any) => {
-  const {index, id} = props
-  const initialValue = props.getValue()
-  // We need to keep and update the state of the cell normally
-  const [value, setValue] = useState(initialValue || '')
+const columnHelper = createColumnHelper<SanityDocument>()
+const SUPPORTED_FIELDS = ['string', 'number', 'boolean']
 
-  // When the input is blurred, we'll call our table meta's updateData function
-  const onBlur = () => {
-    props.table.options.meta?.updateData(index, id, value)
-  }
+type Columns = (
+  | AccessorKeyColumnDef<SanityDocument, unknown>
+  | GroupColumnDef<SanityDocument, unknown>
+)[]
 
-  const handleChange = useCallback(
-    (e: FormEvent<HTMLInputElement>) => setValue(e.currentTarget.value),
-    [],
-  )
+const getColsFromSchemaType = (schemaType: SchemaType, parentalField?: string): Columns => {
+  //@ts-expect-error - wip.
+  return schemaType.fields.reduce((cols: Columns[], field: any) => {
+    const {type, name} = field
+    if (SUPPORTED_FIELDS.includes(type.name)) {
+      const nextCol = columnHelper.accessor(
+        parentalField ? `${parentalField}.${field.name}` : field.name,
+        {
+          header: field.type.title,
+          enableHiding: true,
+          cell: (info) => <SheetListCell {...info} type={type} />,
+        },
+      )
+
+      return [...cols, nextCol]
+    }
 
-  useEffect(() => {
-    setValue(initialValue)
-  }, [initialValue])
+    // if first layer nested object
+    if (type.name === 'object' && !parentalField) {
+      return [
+        ...cols,
+        columnHelper.group({header: name, columns: getColsFromSchemaType(type, field.name)}),
+      ]
+    }
 
-  return <TextInput value={value as string} onChange={handleChange} onBlur={onBlur} />
+    return cols
+  }, [])
 }
 
-const columnHelper = createColumnHelper<SanityDocument>()
-const SUPPORTED_FIELDS = ['string', 'number']
+// Type guard function to check if a column is of type GroupColumnDef
+function isAccessorKeyColumnDef(
+  col: Columns[number],
+): col is AccessorKeyColumnDef<SanityDocument, unknown> {
+  return (col as AccessorKeyColumnDef<SanityDocument, unknown>).accessorKey !== undefined
```

Comment:
Using the [`in` operator to narrow](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) might be better as it doesn't require a cast.

```suggestion
  return 'accessorKey' in col
```

Reply:
- [ ] Use in for type guards